src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -319,20 +319,36 @@ static bool Contains(ROM[] array, ROM item, int count)
             }
         }
 
-        static void Add(Dictionary<string, (GraphQLField Field, FieldType FieldType)> fields, GraphQLField field, FieldType fieldType)
+        static void Add(ExecutionContext context, Dictionary<string, (GraphQLField Field, FieldType FieldType)> fields, GraphQLField field, FieldType fieldType)
         {
             string name = field.Alias != null ? field.Alias.Name.StringValue : fieldType.Name; //ISSUE:allocation in case of alias
 
-            fields[name] = fields.TryGetValue(name, out var original)
-                ? (new GraphQLField(original.Field.Name) // Sets a new field selection node with the child field selection nodes merged with another field's child field selection nodes.
+            if (fields.TryGetValue(name, out var original))
+            {
+                var newField = new GraphQLField(original.Field.Name) // Sets a new field selection node with the child field selection nodes merged with another field's child field selection nodes.
                 {
                     Alias = original.Field.Alias,
                     Arguments = original.Field.Arguments,
                     SelectionSet = Merge(original.Field.SelectionSet, field.SelectionSet),
                     Directives = original.Field.Directives,
                     Location = original.Field.Location,
-                }, original.FieldType)
-                : (field, fieldType);
+                };
+                var argumentValues = context.ArgumentValues;
+                if (argumentValues != null && argumentValues.TryGetValue(original.Field, out var arguments))
+                {
+                    ((IDictionary<GraphQLField, IDictionary<string, ArgumentValue>>)argumentValues)[newField] = arguments;
+                }
+                var directiveValues = context.DirectiveValues;
+                if (directiveValues != null && directiveValues.TryGetValue(original.Field, out var directives))
+                {
+                    ((IDictionary<ASTNode, IDictionary<string, DirectiveInfo>>)directiveValues)[newField] = directives;
+                }
```

Comment:
This code contains the cast.  Essentially, `ExecutionContext.ArgumentValues` and `ExecutionContext.DirectiveValues` must implement `IDictionary<TKey, TValue>` or an `InvalidCastException` will occur if there are selection sets being merged where the parent field has arguments.  In a v9 of GraphQL.NET we should make the above-referenced properties `IDictionary` instead of `IReadOnlyDictionary`.